Pr1

lrumancik opened the pull request:
New eBPF program type to allow filtering of IO at block device layer.

lrumancik committed:
bpf: io_filter attach/detach program to gendisk

lrumancik committed:
bpf: run io_filter program on each submit_bio

orbekk commented on block/Makefile:
```diff
@@ -36,3 +36,6 @@ obj-$(CONFIG_BLK_DEBUG_FS)	+= blk-mq-debugfs.o
 obj-$(CONFIG_BLK_DEBUG_FS_ZONED)+= blk-mq-debugfs-zoned.o
 obj-$(CONFIG_BLK_SED_OPAL)	+= sed-opal.o
 obj-$(CONFIG_BLK_PM)		+= blk-pm.o
+ifeq ($(CONFIG_BPF_IO_FILTER),y)
```
Do you mean CONFIG_BPF_JIT here? The CONFIG_BPF_IO_FILTER looks redundant.

lrumancik replied:
Changed to CONFIG_BPF_JIT

orbekk commented on samples/bpf/bpf_load.c:
```diff
@@ -87,6 +87,7 @@ static int load_and_attach(const char *event, struct bpf_insn *prog, int size)
 	bool is_sockops = strncmp(event, "sockops", 7) == 0;
 	bool is_sk_skb = strncmp(event, "sk_skb", 6) == 0;
 	bool is_sk_msg = strncmp(event, "sk_msg", 6) == 0;
+	bool is_submit_bio = strncmp(event, "submit_bio", 10) == 0;
```
I think sticking with io_filter might be a better event name here. As a user it doesn't feel relevant that the filtering happens inside submit_bio

lrumancik replied:
changed to io_filter for now

orbekk commented on samples/bpf/io_filter_kern.c:
```diff
@@ -0,0 +1,13 @@
+char _license[] SEC("license") = "GPL";
+
+SEC("submit_bio")
+int filter_io(struct bio *b)
```
This is fine for this PR, but we will need to replace struct bio here with something else eventually.

orbekk commented on tools/include/uapi/linux/bpf.h:
```diff
@@ -184,6 +184,7 @@ enum bpf_prog_type {
 	BPF_PROG_TYPE_STRUCT_OPS,
 	BPF_PROG_TYPE_EXT,
 	BPF_PROG_TYPE_LSM,
+	BPF_PROG_TYPE_IO_FILTER,
```
The changes to this file would be appropriate to move to the commit that introduces BPF_PROG_TYPE_IO_FILTER in the kernel.

lrumancik replied:
I believe this got moved to the correct commit during the reorganizing

orbekk commented on samples/bpf/io_filter_user.c:
```diff
@@ -0,0 +1,59 @@
+#define _GNU_SOURCE
```
I don't think we need to keep intermediate versions of this program. Can you create a single commit at the end of your commit chain that contains the final sample program?

lrumancik replied:
Done

orbekk commented on samples/bpf/io_filter_user.c:
```diff
@@ -10,24 +10,43 @@
+	devfd = open(argv[1], O_RDWR | O_NONBLOCK);
+
+	if (devfd == -1) {
+		printf("failed to open lirc device %s: %m\n", argv[1]);
```
lirc -> block

lrumancik replied:
done

orbekk commented on samples/bpf/io_filter_user.c:
```diff
@@ -10,24 +10,43 @@
+
+	printf("attached bpf submit_bio program to device\n");
+
 	strcpy(test_file, getenv("HOME"));
```
The sample program shouldn't (over)write a random file in the home directory. Let's ask the user for a path? (We could use mktemp(), but there is no guarantee that /tmp is on the same device as the one we attach to)

lrumancik replied:
Added test file path as third argument

lrumancik replied:
Could potentially lookup device mount point then use O_TMPFILE in open call to make temporary file on that device but that might be overkill

orbekk commented on block/blk-bpf-io-filter.h:
```diff
@@ -0,0 +1,13 @@
+#ifdef CONFIG_BPF_IO_FILTER
+int io_filter_bpf_run(struct bio *bio);
+#else
+static inline int io_filter_bpf_run(struct bio *bio) { }
```
This needs to return something. It probably doesn't compile if CONFIG_BPF_IO_FILTER is not set.

lrumancik replied:
Added return 0 to allow IO by default

orbekk commented on kernel/bpf/bpf_io_filter.c:
```diff
@@ -211,3 +214,16 @@ int io_filter_prog_detach(const union bpf_attr *attr)
+{
+	int ret = 0;
+
+	if (bio->bi_disk->prog) {
```
Will this require some locking? E.g., program got get detached while this is running

lrumancik replied:
Added rcu locking to assignment of bpf_prog * and to this section, please check

orbekk commented on kernel/bpf/bpf_io_filter.c:
```diff
@@ -146,6 +146,18 @@ static bool io_filter_is_valid_access(int off, int size,
 	return true;
 }
 
+static int io_filter_bpf_attach(struct gendisk *disk, struct bpf_prog *prog)
```
I believe attach/detach can happen concurrently with IOs to the disk. May need some locks here

lrumancik replied:
Added rcu locking

orbekk commented on samples/bpf/io_filter_user.c:
```diff
@@ -10,24 +10,43 @@
+
+	printf("loaded submit_bio program successfully\n");
+
+	devfd = open(argv[1], O_RDWR | O_NONBLOCK);
```
Can we open with O_RDONLY?

lrumancik replied:
Done

orbekk commented on samples/bpf/io_filter_user.c:
```diff
@@ -10,24 +10,43 @@
 	strcat(test_file, "/submit_bio_test");
 
-	fd = open(test_file, O_WRONLY|O_CREAT|O_TRUNC, S_IRWXU | S_IRWXG | S_IRWXO);
+	fd = open(test_file, O_WRONLY|O_CREAT|O_TRUNC, S_IRWXU|S_IRWXG|S_IRWXO);
```
Can we drop the mode argument?

lrumancik replied:
Open manpage says the mode arguments are required when O_CREAT is used. However, it does indeed work (can write to file) when mode argument is dropped, but looking at file after running, permissions are "----------"

orbekk commented on block/Makefile:
```diff
@@ -36,3 +36,6 @@ obj-$(CONFIG_BLK_DEBUG_FS)	+= blk-mq-debugfs.o
 obj-$(CONFIG_BLK_DEBUG_FS_ZONED)+= blk-mq-debugfs-zoned.o
```
This 'move io filter' commit could get folded into the commit that introduced these files.

orbekk reviewed (changes requested):
@lrumancik: This looks great so far. I would have liked an overall description (and a better name) to explain the scope of these changes, and perhaps some output from running the sample program.

@harshadjs @srnym: Please take a look.

srnym commented on block/blk-core.c:
```diff
@@ -927,6 +929,12 @@ generic_make_request_checks(struct bio *bio)
+	printk(KERN_DEBUG "io_filter_bpf_run returned %d", ret);
+	/*
+	if (io_filter_bpf_run(bio))
+		goto end_io;
```
goto end_io makes sense here

harshadjs replied:
+1

lrumancik replied:
Changed from printing value to using the if statement

srnym reviewed (commented):
Looks good to me! Agree with KJs comments and its understandable that there are few places not ironed out(Like dependency on trace_helpers) because its just the first set of the patches.

harshadjs commented on block/bpf_io_filter.c:
```diff
@@ -0,0 +1,229 @@
+	f = fdget(attr->target_fd);
+	disk  = I_BDEV(f.file->f_mapping->host)->bd_disk;
+
+	ret = io_filter_bpf_attach(disk, prog);
```
What if a program is already attached? We need some kind of synchronization here, I think networking uses rcu for this, if I'm not wrong.

https://lwn.net/Articles/262464/

lrumancik replied:
Added TODO in code to change program to array of programs to allow multiple programs to be attached to device

harshadjs commented on block/bpf_io_filter.c:
```diff
@@ -0,0 +1,229 @@
+		return -EINVAL;
+
+	f = fdget(attr->target_fd);
+	disk  = I_BDEV(f.file->f_mapping->host)->bd_disk;
```
I'm not very sure about this, but there can be a block device that doesn't have any attached gendisk. If we don't check if "disk" is valid before using it, we may end up crashing the kernel.

lrumancik replied:
Do you have any suggestions on how to error check this? fdget and I_BDEV seem to be just accessing fields / converting pointers / etc, not clearly returning errors when something may be wrong

lrumancik replied:
Done

harshadjs commented on block/bpf_io_filter.c:
```diff
@@ -0,0 +1,229 @@
+	return true;
+}
+
+static int io_filter_bpf_attach(struct gendisk *disk, struct bpf_prog *prog)
```
I'm not sure I understand why bpf_attach/detach is separate from prog_attach/detach?

lrumancik replied:
I was following the structure of other bpf program types, but considering how short the bpf_attach/detach is, it may be cleaner to collapse it into just the prog_attach/detach

lrumancik replied:
removed separate bpf attach/detach functions

harshadjs commented on block/bpf_io_filter.c:
```diff
@@ -0,0 +1,229 @@
+	f = fdget(attr->target_fd);
+	disk  = I_BDEV(f.file->f_mapping->host)->bd_disk;
+
+	ret = io_filter_bpf_detach(disk, prog);
```
Similar synchronization comment here, we need it here as well. We may end up detaching a program which is running.

lrumancik replied:
Added locking

harshadjs commented on include/uapi/linux/bpf.h:
```diff
@@ -184,6 +184,7 @@ enum bpf_prog_type {
 	BPF_PROG_TYPE_STRUCT_OPS,
 	BPF_PROG_TYPE_EXT,
 	BPF_PROG_TYPE_LSM,
+	BPF_PROG_TYPE_IO_FILTER,
```
Maybe all these things also need to be protected under #ifdef?

lrumancik replied:
Hm I am not sure, are you talking about an #ifdef on a Kconfig option?

orbekk replied:
IIUC this is the userspace header so it should be here unconditionally -- if the kernel doesn't have io_filter enabled then it will just refuse to attach anything

harshadjs replied:
ah that's right.

harshadjs commented on block/bpf_io_filter.c:
```diff
@@ -0,0 +1,229 @@
+	if (err)
+		goto events_exit;
+
+	progs_kset = kset_create_and_add("progs", NULL, bpf_io_filter_kobj);
```
How are we planning to use this?

lrumancik replied:
I am not quite sure, I did not have to modify this at all. Any ideas @srnym ?

srnym replied:
I dont recall the need for this.

lrumancik replied:
Removed init function and everything is still working fine, can add back in later if needed

lrumancik committed:
bpf: add io_filter sample code

lrumancik committed:
bpf: add new prog_type BPF_PROG_TYPE_IO_FILTER

Introducing a new program type BPF_PROG_TYPE_IO_FILTER and
a new attach type BPF_BIO_SUBMIT.

This program type is intended to help filter block io requests
and possibly drop writes to disks.

Co-authored-by: Author: Saranya Muruganandam <saranyamohan@google.com>

orbekk commented on block/bpf_io_filter.c:
```diff
@@ -0,0 +1,172 @@
+}
+*/
+
+const struct bpf_prog_ops io_filter_prog_ops = {
```
Should these structs be static? @harshadjs @srnym

harshadjs replied:
I think this is okay, if this variable isn't declared in any header file, other files won't have a way to access it anyway.

orbekk replied:
Isn't this why you would make it static, so it won't collide with other variables? Not sure what the kernel style says about this

harshadjs replied:
Yeah, sorry my bad. We do need to make it static to avoid collision. Such structs are often declared static to avoid the collision problem.

harshadjs reviewed (approved):
Looks good to me!
